File: cpp/algorithms-dsa/leetcode/arrays-hashing/contains_duplicate.cpp
/**
 * LeetCode 217: Contains Duplicate
 * Topic: Arrays & Hashing
 * Difficulty: Easy
 *
 * Return true if any value appears at least twice.
 * Time: O(n), Space: O(n)
 */
#include <iostream>
#include <vector>
#include <unordered_set>
#include <cassert>
using namespace std;

bool containsDuplicate(vector<int>& nums) {
    unordered_set<int> seen;
    for (int n : nums) {
        if (seen.count(n)) return true;
        seen.insert(n);
    }
    return false;
}

int main() {
    vector<int> v1 = {1, 2, 3, 1};
    assert(containsDuplicate(v1) == true);

    vector<int> v2 = {1, 2, 3, 4};
    assert(containsDuplicate(v2) == false);

    vector<int> v3 = {1, 1, 1, 3, 3, 4, 3, 2, 4, 2};
    assert(containsDuplicate(v3) == true);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/go/algorithms-dsa/leetcode/arrays_hashing/contains_duplicate.go
// LeetCode 217: Contains Duplicate
// Given an integer array nums, return true if any value appears at least twice.
// Time: O(n), Space: O(n)
package main

import "fmt"

func containsDuplicate(nums []int) bool {
	seen := make(map[int]bool)
	for _, num := range nums {
		if seen[num] {
			return true
		}
		seen[num] = true
	}
	return false
}

func assertBool(actual, expected bool, msg string) {
	if actual != expected {
		panic(fmt.Sprintf("FAIL %s: got %v, want %v", msg, actual, expected))
	}
	fmt.Printf("PASS: %s\n", msg)
}

func main() {
	assertBool(containsDuplicate([]int{1, 2, 3, 1}), true, "has duplicate")
	assertBool(containsDuplicate([]int{1, 2, 3, 4}), false, "no duplicate")
	assertBool(containsDuplicate([]int{1, 1, 1, 3, 3, 4, 3, 2, 4, 2}), true, "many duplicates")
	assertBool(containsDuplicate([]int{}), false, "empty array")

	fmt.Println("All tests passed!")
}
